server/src-lib/Data/Time/Clock/Units.hs

```diff
@@ -25,56 +25,72 @@ You can also go the other way using the constructors rather than the selectors:
 0.5
 @
 
+The 'Read' instances for these types mirror the behavior of the 'RealFrac' instance wrt numeric
+literals for convenient serialization (e.g. when working with env vars):
+
+@
+>>> read "1.2" :: Milliseconds 
+Milliseconds {milliseconds = 0.0012s}
+@
+
 Generally, it doesn’t make sense to pass these wrappers around or put them inside data structures,
-since any function that needs a duration should just accept a 'DiffTime', but they’re useful for
-literals and conversions to/from other types. -}
+since any function that needs a duration should just accept a 'NominalDiffTime', but they’re useful
+for literals and conversions to/from other types. -}
 module Data.Time.Clock.Units
   ( Days(..)
   , Hours(..)
   , Minutes(..)
-  , Seconds
-  , seconds
+  , Seconds(..)
   , Milliseconds(..)
   , Microseconds(..)
   , Nanoseconds(..)
+  -- * Reexports
+  -- | We use 'NominalDiffTime' as the standard type for duration in our code.
+  --
+  -- NOTE: some care must be taken especially when 'NominalDiffTime' interacts with 'UTCTime':
+  --
+  --  - 'addUTCTime' and 'diffUTCTime' do not attempt to handle leap seconds
+  --  - a 'NominalDiffTime' my be logically negative
+  , NominalDiffTime
   ) where
 
 import           Prelude
 
+import           Control.Arrow   (first)
 import           Data.Proxy
 import           Data.Time.Clock
 import           GHC.TypeLits
+import           Numeric         (readFloat)
 
-type Seconds = DiffTime
-
-seconds :: DiffTime -> DiffTime
-seconds = id
+newtype Seconds = Seconds { seconds :: NominalDiffTime }
+  deriving (Show, Eq, Ord)
+  deriving (Read, Num, Fractional, Real, RealFrac) via (TimeUnit (SecondsP 1))
 
-newtype Days = Days { days :: DiffTime }
+newtype Days = Days { days :: NominalDiffTime }
   deriving (Show, Eq, Ord)
-  deriving (Num, Fractional, Real, RealFrac) via (TimeUnit (SecondsP 86400))
+  deriving (Read, Num, Fractional, Real, RealFrac) via (TimeUnit (SecondsP 86400))
```

Comment:
Do we really need the `Read` instances here? If so, why?

Reply:
It was motivated from an application in this refactoring of a function like `readEnv :: Read a=> String -> IO a` (I can find the actually code if you need), but it seems generally useful for this kind of thing

Reply:
I don’t really care for `Read` as a typeclass, but I think this instance is especially sketchy, since it breaks the informal “law” that `Show` and `Read` instances are intended to roundtrip, if both are defined. This is not the case here, since we are deriving the stock `Show` instance but the underlying `Read` instance.

I’d prefer some slightly more explicit parsing logic. What about deriving [`ParseTime`](https://hackage.haskell.org/package/time-1.9.3/docs/Data-Time-Format.html#t:ParseTime), instead?

Reply:
I'm fine removing the `Read` instance or implementing a `Show` that works in the same way

Reply:
I’d rather leave the `Read` instance off if there isn’t any code using it right now.

Reply:
(as I said there is some code using it right now that motivated the instances, but that can be rewritten)

Reply:
Ah, I misunderstood if you were saying it was currently using it or if it was previously using it. If it’s currently using it, I won’t die on this hill; I’ll leave it up to you whether you want to change it or not.

The reason I don’t like `Read` is that I never have any idea what it does for any given type. Does it try to parse a Haskell expression that would produce the value? Does it use some custom serialization format? In the case of `readFloat` and other `Read` operations on numeric types, it goes and actually tries to parse a Haskell number literal, which I think is rarely what people intend!

Admittedly, in this particular instance, it probably isn’t so bad, so maybe it’s okay. I’m just always wary of relying on `Read`, since different instances work in wildly different ways.